### PR TITLE
Shorten `rustc_middle::ty::mod`

### PR DESCRIPTION
**Pull request by Nicholas-Baron**
Related to #60302. 

This PR moves all `Adt*`, `Assoc*`, `Generic*`, and `UpVar*` types to separate files.
This, alongside some `use` reordering, puts `mod.rs` at ~2,200 lines, thus removing the `// ignore-tidy-filelength`.

The particular groups were chosen as they had 4 or more "substantive" members.

### CODE REVIEW

**Comment by petrochenkov** [comment]
Not sure who maintains this code now.
r? @jackh726 @nikomatsakis

**Comment by cjgillot** [comment]
What is the status of the work to move the type definitions to `rustc_type_ir`?

**Comment by jackh726** [comment]
> What is the status of the work to move the type definitions to `rustc_type_ir`?

Somewhat blocked on figuring out the `Encodable`/`Decodable` situation. @detrumi and @zaharidichev have been doing [here](https://github.com/rust-lang/rust/compare/master...zaharidichev:move-encodable-to-type-ir)

File: compiler/rustc_middle/src/ty/upvar.rs
Comment (jackh726) [inline compiler/rustc_middle/src/ty/upvar.rs:1]: This is the only file that I'm not sure holds it weight. I feel like this should also include some of the min capture related stuff or just things related to closure captures in general.

**Review by jackh726 — CHANGES_REQUESTED** [review]
I big 👍 on me for the PR in general. For `upvar.rs`, I would like that to either be reverted or to add some other closure capture related things and changed to `closure_capture` or something like that :)

File: compiler/rustc_middle/src/ty/mod.rs
Comment (jackh726) [inline compiler/rustc_middle/src/ty/mod.rs:407]: Maybe more this to `closure.rs` too?

**Review by jackh726 — COMMENTED** [review]
This looks better. I think anything related directly to closure captures can go in `closure.rs`

File: compiler/rustc_middle/src/ty/mod.rs
Comment (Nicholas-Baron) [reply compiler/rustc_middle/src/ty/mod.rs]: Moved that and some other types that are only referenced in `closure.rs`.

**Comment by jackh726** [comment]
Looks good to me! r=me when CI is green

**Comment by Nicholas-Baron** [comment]
> Looks good to me! r=me when CI is green

So when the CI returns all good, I write `r=jackh726` to bors?

**Comment by jackh726** [comment]
@bors delegate+

Yes :)

**Comment by Nicholas-Baron** [comment]
I hope I did not mess up anything.

**Comment by Mark-Simulacrum** [comment]
This seems to have been a pretty big (~30%) [regression](https://perf.rust-lang.org/compare.html?start=066f01d81bfbed746f6b6cf27a0426d829e8e832&end=b3ac52646f7591a811fa9bf55995b24fd17ece08&stat=instructions:u) on diesel-doc builds, presumably due to inlining changes - self-profile data in particular points at the evaluate_obligation query. It's possible we can just leave this, but might be good to add a couple inline attributes which should resolve the regression.

**Comment by Nicholas-Baron** [comment]
> This seems to have been a pretty big (~30%) [regression](https://perf.rust-lang.org/compare.html?start=066f01d81bfbed746f6b6cf27a0426d829e8e832&end=b3ac52646f7591a811fa9bf55995b24fd17ece08&stat=instructions:u) on diesel-doc builds, presumably due to inlining changes - self-profile data in particular points at the evaluate_obligation query. It's possible we can just leave this, but might be good to add a couple inline attributes which should resolve the regression.

Are there any particular functions which would benefit from inlining?

**Comment by Mark-Simulacrum** [comment]
It's hard to say without more investigation; probably a cachegrind or similar tooling might give a hint. I unfortunately don't have time to write up some directions, but a start would be to:

* checkout the rust compiler tree at the before & after commits around this PR's merge and build locally to get debuginfo
* run cachegrind on each with the diesel doc benchmark (either via the perf collector, some docs [here](https://github.com/rust-lang/rustc-perf/tree/master/collector#how-to-benchmark-a-change-on-your-own-machine), or more directly)
* look to see if any small functions stopped getting inlined

**Comment by Nicholas-Baron** [comment]
@Mark-Simulacrum I tried to use the `rustc-perf/collector` locally. However, linking LLVM with `release-debuginfo = true` leads to OOM for my 16 GB dev machine. I never got to building even stage0.

**Comment by Mark-Simulacrum** [comment]
Oh, you almost certainly only need debug info for the compiler, not llvm

**Comment by Nicholas-Baron** [comment]
@Mark-Simulacrum I have run into an issue which I have filed [here](https://github.com/rust-lang/rustc-perf/issues/854). It looks like I am roadblocked for the current moment.

**Comment by Nicholas-Baron** [comment]
@Mark-Simulacrum I guess I am leaving this alone. There are probably larger performance gains to be had elsewhere in the compiler.

**Pull request merged**
